File: app12_kotlin_android/app/src/main/assets/code_samples/java/sql-injection-prevention_defense.java
// SECURE pattern
String name = request.getParameter("name");
// WHY: PreparedStatement sends the query shape and the value separately, so user input can never alter the query structure
PreparedStatement stmt = connection.prepareStatement("SELECT id, email FROM users WHERE name = ?");
stmt.setString(1, name);
ResultSet rs = stmt.executeQuery();

File: app11_swift_ios/SwiftGuardApp/Resources/CodeSamples/python/sql-injection-prevention_defense.py
# SECURE pattern
from django.db import connection

def find_user_by_name(request):
    name = request.GET["name"]
    with connection.cursor() as cursor:
        # WHY: %s is a placeholder bound by the DB driver, never concatenated into the SQL text
        cursor.execute("SELECT id, email FROM users WHERE name = %s", [name])
        return cursor.fetchall()
